src/checkpoint.hh

```diff
@@ -0,0 +1,440 @@
+       bool table_supported = false;
+
+       for (Type::Type supported_type : SUPPORTED_TYPES) {
+         if (table_type == supported_type) {
+           table_supported = true;
+           break;
+         }
+       }
+
+       if (!table_supported) {
+         answer_type_supported = false;
+         break;
+       }
+     }
+
+     return answer_type_supported;
+  }
+
+  void include(Printer::Base &stream) {
+     stream << "#define CHECKPOINTING_INTEGRATED" << endl << endl;
+     stream << "extern \"C\" {" << endl;
+     stream << indent() << "#include <unistd.h>" << endl;
+     stream << "}" << endl;
+     stream << "#include \"boost/serialization/vector.hpp\"" << endl;
+     stream << "#include \"boost/archive/binary_iarchive.hpp\"" << endl;
+     stream << "#include \"boost/archive/binary_oarchive.hpp\"" << endl;
+     stream << "#include \"boost/dll.hpp\"" << endl;
+     stream << "#include \"boost/filesystem.hpp\"" << endl;
+     stream << "#include <atomic>" << endl;
+     stream << "#include <filesystem>" << endl;
+     stream << "#include <thread>" << endl << endl;
+  }
+
+  void archive(Printer::Base &stream) {
+     inc_indent(); inc_indent();
+     stream << indent();
+     stream << "void archive(const std::string &tname) {" << endl;
+     inc_indent();
+     stream << indent() << "// save the DP table/array to disk" << endl;
+     stream << indent() << "try {" << endl;
+     inc_indent();
+     stream << indent() << "std::ofstream array_fout(table_path, "
+                            "std::ios::binary);" << endl;
+     stream << indent() << "if (!(array_fout.good())) {" << endl;
+     stream << indent() << "  throw std::ofstream::failure(\"\");" << endl;
+     stream << indent() << "}" << endl;
+     stream << indent() << "boost::archive::binary_oarchive "
+                            "array_out(array_fout);" << endl;
+     stream << indent() << "array_out << array;" << endl;
+     stream << indent() << "array_fout.close();" << endl;
+     stream << indent() << "std::cout << \"Info: Archived \\\"\" << tname << "
+            << "\"\\\" table at current checkpoint.\" << std::endl;" << endl;
+     dec_indent();
+     stream << indent() << "} catch (const std::ofstream::failure &e) {"
+             << endl;
+     stream << indent() << "  std::cerr << \"Couldn't create table archive "
+             << "at path \" << table_path << \".\\n\"" << endl;
+     stream << indent() << "            << \"Please ensure that the directory "
+             << "exists and that you have write permissions "
+             << "for this directory.\\n\";" << endl;
+     stream << indent() << "} catch (const std::exception &e) {" << endl;
+     inc_indent();
+     stream << indent() << "std::cerr << \"Error trying to archive \\\"\" "
+                            "<< tname << \"\\\" table.\""
+             << endl;
+     stream << indent() << "          << \" Will retry in \" "
+            << "<< formatted_interval << \".\n\";" << endl;
```

Comment:
`\n` leads to:
```
out.hh:89:65: warning: missing terminating " character
   89 |                   << " Will retry in " << formatted_interval << ".
      |                                                                 ^
out.hh:89:65: error: missing terminating " character
   89 |                   << " Will retry in " << formatted_interval << ".
      |                                                                 ^~
```

Reply:
Yeah, I just noticed that as well. All those escape characters confuse me sometimes.